stdlib/string.gr

```diff
@@ -1,72 +1,522 @@
-export *
+import WasmI32 from "runtime/unsafe/wasmi32"
+import Memory from "runtime/unsafe/memory"
+import Errors from "runtime/unsafe/errors"
+import { tagSimpleNumber, allocateArray, allocateChar, allocateString } from "runtime/dataStructures"
+import foreign wasm throwError: (WasmI32, WasmI32, WasmI32) -> a from "grainRuntime"
 
 // Concatenate two strings.
 // @param str1: String - The first string
 // @param str2: String - The second string
 // @returns String
-let concat = (++)
+export let concat = (++)
 
 // Get the character length of a string.
 // @param input: String - The string to check
 // @returns Number
-import foreign wasm length : String -> Number from "stdlib-external/string"
+@disableGC
+export let length = (s: String) => {
+  let s = WasmI32.fromGrain(s)
+  let size = WasmI32.load(s, 4n)
+
+  let mut len = 0n
+  let mut ptr = WasmI32.add(s, 8n)
+  let end = WasmI32.add(ptr, size)
+
+  while (WasmI32.ltU(ptr, end)) {
+    let byte = WasmI32.load8U(ptr, 0n)
+    if (WasmI32.ne(WasmI32.and(byte, 0xC0n), 0x80n)) {
+      len = WasmI32.add(len, 1n);
+    }
+    ptr = WasmI32.add(ptr, 1n);
+  }
+
+  tagSimpleNumber(len)
+}
 
 // Get the byte length of a string.
 // @param input: String - The string to check
 // @returns Number
-import foreign wasm byteLength : String -> Number from "stdlib-external/string"
+@disableGC
+export let byteLength = (s: String) => {
+  let s = WasmI32.fromGrain(s)
+  tagSimpleNumber(WasmI32.load(s, 4n))
+}
 
-import foreign wasm indexOf : (String, String) -> Number from "stdlib-external/string"
 // Find the start index of a substring.
 // @param sub: String - The substring to find
 // @param input: String - The string to check
 // @returns Option<Number>
-let indexOf = (p, s) => {
-  let result = indexOf(p, s)
-  if (result < 0) {
-    None
+@disableGC
+export let indexOf = (p: String, s: String) => {
+  let p = WasmI32.fromGrain(p)
+  let s = WasmI32.fromGrain(s)
+  
+  let size = WasmI32.load(s, 4n)
+  let psize = WasmI32.load(p, 4n)
+
+  let (>) = WasmI32.gtU
+  let (<) = WasmI32.ltU
+  let (==) = WasmI32.eq
+  let (+) = WasmI32.add
+  let (-) = WasmI32.sub
+  let (&) = WasmI32.and
+
+  if (psize > size) {
+    let none = None
+    Memory.incRef(WasmI32.fromGrain(none))
+    none
   } else {
-    Some(result)
+    let mut idx = 0n
+    let mut ptr = s + 8n
+    let mut pptr = p + 8n
+    let end = ptr + size - psize + 1n
+
+    let mut result = -1n
+
+    while (ptr < end) {
+      if (Memory.compare(ptr, pptr, psize) == 0n) {
+        result = idx
+        break
+      }
+      idx += 1n
+      let byte = WasmI32.load8U(ptr, 0n)
+      if ((byte & 0x80n) == 0x00n) {
+        ptr += 1n;
+      } else if ((byte & 0xF0n) == 0xF0n) {
+        ptr += 4n;
+      } else if ((byte & 0xE0n) == 0xE0n) {
+        ptr += 3n;
+      } else {
+        ptr += 2n;
+      }
+    }
+
+    if (result == -1n) {
+      let none = None
+      Memory.incRef(WasmI32.fromGrain(none))
```

Comment:
Why do you incRef `none` but not the `Some`?

Reply:
This just maintains GC semantics. A function is expected to return a value with a reference count of +1. Because of this rule, if I call a function and want to just return the result, I don't need to incRef it because its count is +1. If I've received a reference to something, like a parameter, its count is +0, so when I return it it needs an incRef to have a count of +1. (Writing `None` in Grain doesn't do an allocation, it just references the value None.)

Much easier to let the compiler figure it out for you 😛